### PR TITLE
Refactor iterator APIs to be on parity with the latest spec

### PR DESCRIPTION
**Pull request by jedel1043**
Alternative name: The Great Refactoring Of Iterators

This pulls the latest changes from the spec around iterators, which both makes the code easier in some parts and also future-proofs for any new spec additions that use iterator APIs.

### CODE REVIEW

File: core/engine/src/builtins/promise/mod.rs [hunk 1]
```diff
@@ -1100,19 +1058,40 @@ impl Promise {
...
+        //     i. Set remainingElementsCount.[[Value]] to remainingElementsCount.[[Value]] - 1.
+        remaining_elements_count.set(remaining_elements_count.get() - 1);
+
+        //     ii. If remainingElementsCount.[[Value]] = 0, then
+        if remaining_elements_count.get() == 0 {
+            //         1. Let valuesArray be CreateArrayFromList(values).
+            let values_array =
+                Array::create_array_from_list(values.borrow().as_slice().iter().cloned(), context);
+
+            //         2. Perform ? Call(resultCapability.[[Resolve]], undefined, « valuesArray »).
```
Comment (raskad) [inline core/engine/src/builtins/promise/mod.rs:1085]:
```suggestion
            // 1. Let valuesArray be CreateArrayFromList(values).
            let values_array =
                Array::create_array_from_list(values.borrow().as_slice().iter().cloned(), context);

            // 2. Perform ? Call(resultCapability.[[Resolve]], undefined, « valuesArray »).
```

File: core/engine/src/builtins/promise/mod.rs [hunk 2]
```diff
@@ -1349,9 +1292,33 @@ impl Promise {
...
+                    .iter()
+                    .cloned()
+                    .map(JsError::from_opaque)
+                    .collect(),
+            )
+            .with_message("no promise in Promise.any was fulfilled.");
+
+            //             2. Perform ! DefinePropertyOrThrow(error, "errors", PropertyDescriptor { [[Configurable]]: true, [[Enumerable]]: false, [[Writable]]: true, [[Value]]: CreateArrayFromList(errors) }).
+            //             3. Return ThrowCompletion(error).
+            return Err(error.into());
```
Comment (raskad) [inline core/engine/src/builtins/promise/mod.rs:1317]:
```suggestion
            // 1. Let error be a newly created AggregateError object.
            let error = JsNativeError::aggregate(
                errors
                    .borrow()
                    .iter()
                    .cloned()
                    .map(JsError::from_opaque)
                    .collect(),
            )
            .with_message("no promise in Promise.any was fulfilled.");

            // 2. Perform ! DefinePropertyOrThrow(error, "errors", PropertyDescriptor { [[Configurable]]: true, [[Enumerable]]: false, [[Writable]]: true, [[Value]]: CreateArrayFromList(errors) }).
            // 3. Return ThrowCompletion(error).
            return Err(error.into());
```

**Review by raskad — APPROVED** [review]
Nice work!

**Review by nekevss — APPROVED** [review]
+1 for alternative name 😄 

Looks great!

**Pull request merged**